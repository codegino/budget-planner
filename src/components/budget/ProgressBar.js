import React from 'react';
import { Text, View, StyleSheet } from 'react-native';
import colors from '../../constants/colors';

const styles = StyleSheet.create({
  container: {
    borderColor: colors.primaryLight,
    backgroundColor: colors.primaryDark,
    borderWidth: 2,
    borderRadius: 5,
    padding: 1,
    overflow: 'hidden',
    justifyContent: 'center',
    width: '100%',
  },
  progress: {
    minHeight: 20,
    borderRadius: 3,
    flexDirection: 'row',
    alignItems: 'center',
  },
  label: {
    fontWeight: 'bold',
    alignSelf: 'center',
    paddingHorizontal: 5,
  },
  labelWrapper: {
    position: 'absolute',
    alignItems: 'center',
    justifyContent: 'center',
    width: '100%',
  },
});

type ProgressBarProps = {
  numerator: number,
  denominator: number,
}

const ProgressBar = (props: ProgressBarProps) => {
  const progress = (props.numerator / props.denominator) * 100;

  let color = colors.success;
  let textColor = 'white';

  if (progress >= 90) {
    color = colors.danger;
    textColor = colors.darkGray3;
  } else if (progress >= 70) {
    color = colors.warning;
    textColor = colors.darkGray2;
  } else if (progress >= 50) {
    color = colors.notify;
    textColor = colors.darkGray2;
  }

  return (
    <View style={styles.container}>
      <View style={[styles.progress, { width: `${progress}%`, backgroundColor: color }]} />
      <View style={styles.labelWrapper}>
        <Text style={[styles.label, { color: textColor }]}>
          {props.numerator}/{props.denominator}
        </Text>
      </View>
    </View>
  );
};

export default ProgressBar;
