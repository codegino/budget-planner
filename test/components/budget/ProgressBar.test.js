import React from 'react';
import ProgressBar from 'components/budget/ProgressBar';
import renderer from 'react-test-renderer';

describe('renders with color success progress bar when percentage is below 50', () => {
  it('will have an 0% at minimum', () => {
    const greenProgressBarMin = <ProgressBar numerator={0} denominator={100} />;
    expect(renderer.create(greenProgressBarMin).toJSON()).toMatchSnapshot();
  });

  it('will have an 49% at maximum', () => {
    const greenProgressBarMax = <ProgressBar numerator={49} denominator={100} />;
    expect(renderer.create(greenProgressBarMax).toJSON()).toMatchSnapshot();
  });
});


describe('renders with color notify progress bar when percentage is 50-69', () => {
  it('will change color at 50%', () => {
    const greenProgressBarMin = <ProgressBar numerator={50} denominator={100} />;
    expect(renderer.create(greenProgressBarMin).toJSON()).toMatchSnapshot();
  });

  it('will have an 69% at maximum', () => {
    const greenProgressBarMax = <ProgressBar numerator={69} denominator={100} />;
    expect(renderer.create(greenProgressBarMax).toJSON()).toMatchSnapshot();
  });
});

describe('renders with color warning progress bar when percentage is 70-89', () => {
  it('will change color at 70%', () => {
    const greenProgressBarMin = <ProgressBar numerator={70} denominator={100} />;
    expect(renderer.create(greenProgressBarMin).toJSON()).toMatchSnapshot();
  });

  it('will have an 89% at maximum', () => {
    const greenProgressBarMax = <ProgressBar numerator={89} denominator={100} />;
    expect(renderer.create(greenProgressBarMax).toJSON()).toMatchSnapshot();
  });
});

describe('renders with color danger progress bar when percentage 90 and above', () => {
  it('will change color at 90%', () => {
    const greenProgressBarMin = <ProgressBar numerator={90} denominator={100} />;
    expect(renderer.create(greenProgressBarMin).toJSON()).toMatchSnapshot();
  });

  it('will still be color danger after exceeding', () => {
    const greenProgressBarMax = <ProgressBar numerator={1000} denominator={100} />;
    expect(renderer.create(greenProgressBarMax).toJSON()).toMatchSnapshot();
  });
});
