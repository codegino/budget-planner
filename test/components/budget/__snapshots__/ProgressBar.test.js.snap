// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with color danger progress bar when percentage 90 and above will change color at 90% 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#193e2c",
      "borderColor": "#5fbd98",
      "borderRadius": 5,
      "borderWidth": 2,
      "justifyContent": "center",
      "overflow": "hidden",
      "padding": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 3,
          "flexDirection": "row",
          "minHeight": 20,
        },
        Object {
          "backgroundColor": "red",
          "width": "90%",
        },
      ]
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "fontWeight": "bold",
            "paddingHorizontal": 5,
          },
          Object {
            "color": "#555",
          },
        ]
      }
    >
      90
      /
      100
    </Text>
  </View>
</View>
`;

exports[`renders with color danger progress bar when percentage 90 and above will still be color danger after exceeding 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#193e2c",
      "borderColor": "#5fbd98",
      "borderRadius": 5,
      "borderWidth": 2,
      "justifyContent": "center",
      "overflow": "hidden",
      "padding": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 3,
          "flexDirection": "row",
          "minHeight": 20,
        },
        Object {
          "backgroundColor": "red",
          "width": "1000%",
        },
      ]
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "fontWeight": "bold",
            "paddingHorizontal": 5,
          },
          Object {
            "color": "#555",
          },
        ]
      }
    >
      1000
      /
      100
    </Text>
  </View>
</View>
`;

exports[`renders with color notify progress bar when percentage is 50-69 will change color at 50% 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#193e2c",
      "borderColor": "#5fbd98",
      "borderRadius": 5,
      "borderWidth": 2,
      "justifyContent": "center",
      "overflow": "hidden",
      "padding": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 3,
          "flexDirection": "row",
          "minHeight": 20,
        },
        Object {
          "backgroundColor": "yellow",
          "width": "50%",
        },
      ]
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "fontWeight": "bold",
            "paddingHorizontal": 5,
          },
          Object {
            "color": "#333",
          },
        ]
      }
    >
      50
      /
      100
    </Text>
  </View>
</View>
`;

exports[`renders with color notify progress bar when percentage is 50-69 will have an 69% at maximum 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#193e2c",
      "borderColor": "#5fbd98",
      "borderRadius": 5,
      "borderWidth": 2,
      "justifyContent": "center",
      "overflow": "hidden",
      "padding": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 3,
          "flexDirection": "row",
          "minHeight": 20,
        },
        Object {
          "backgroundColor": "yellow",
          "width": "69%",
        },
      ]
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "fontWeight": "bold",
            "paddingHorizontal": 5,
          },
          Object {
            "color": "#333",
          },
        ]
      }
    >
      69
      /
      100
    </Text>
  </View>
</View>
`;

exports[`renders with color success progress bar when percentage is below 50 will have an 0% at minimum 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#193e2c",
      "borderColor": "#5fbd98",
      "borderRadius": 5,
      "borderWidth": 2,
      "justifyContent": "center",
      "overflow": "hidden",
      "padding": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 3,
          "flexDirection": "row",
          "minHeight": 20,
        },
        Object {
          "backgroundColor": "green",
          "width": "0%",
        },
      ]
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "fontWeight": "bold",
            "paddingHorizontal": 5,
          },
          Object {
            "color": "white",
          },
        ]
      }
    >
      0
      /
      100
    </Text>
  </View>
</View>
`;

exports[`renders with color success progress bar when percentage is below 50 will have an 49% at maximum 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#193e2c",
      "borderColor": "#5fbd98",
      "borderRadius": 5,
      "borderWidth": 2,
      "justifyContent": "center",
      "overflow": "hidden",
      "padding": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 3,
          "flexDirection": "row",
          "minHeight": 20,
        },
        Object {
          "backgroundColor": "green",
          "width": "49%",
        },
      ]
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "fontWeight": "bold",
            "paddingHorizontal": 5,
          },
          Object {
            "color": "white",
          },
        ]
      }
    >
      49
      /
      100
    </Text>
  </View>
</View>
`;

exports[`renders with color warning progress bar when percentage is 70-89 will change color at 70% 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#193e2c",
      "borderColor": "#5fbd98",
      "borderRadius": 5,
      "borderWidth": 2,
      "justifyContent": "center",
      "overflow": "hidden",
      "padding": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 3,
          "flexDirection": "row",
          "minHeight": 20,
        },
        Object {
          "backgroundColor": "orange",
          "width": "70%",
        },
      ]
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "fontWeight": "bold",
            "paddingHorizontal": 5,
          },
          Object {
            "color": "#333",
          },
        ]
      }
    >
      70
      /
      100
    </Text>
  </View>
</View>
`;

exports[`renders with color warning progress bar when percentage is 70-89 will have an 89% at maximum 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#193e2c",
      "borderColor": "#5fbd98",
      "borderRadius": 5,
      "borderWidth": 2,
      "justifyContent": "center",
      "overflow": "hidden",
      "padding": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 3,
          "flexDirection": "row",
          "minHeight": 20,
        },
        Object {
          "backgroundColor": "orange",
          "width": "89%",
        },
      ]
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "fontWeight": "bold",
            "paddingHorizontal": 5,
          },
          Object {
            "color": "#333",
          },
        ]
      }
    >
      89
      /
      100
    </Text>
  </View>
</View>
`;
